iam-login-service/src/main/java/it/infn/mw/iam/api/account_linking/DefaultAccountLinkingService.java

```diff
@@ -140,56 +142,49 @@ public void linkX509Certificate(Principal authenticatedUser,
 
     IamAccount userAccount = findAccount(authenticatedUser);
 
-    iamAccountRepository.findByCertificateSubject(x509Credential.getSubject())
-      .ifPresent(linkedAccount -> {
-        if (!linkedAccount.getUuid().equals(userAccount.getUuid())) {
-          throw new AccountAlreadyLinkedError(
-              format("X.509 credential with subject '%s' is already linked to another user",
-                  x509Credential.getSubject()));
-        }
-      });
+    Optional<IamAccount> linkedAccount =
+        certificateRepository.findBySubject(x509Credential.getSubject()).stream().findFirst();
 
-    Optional<IamX509Certificate> linkedCert = userAccount.getX509Certificates()
-      .stream()
-      .filter(c -> c.getSubjectDn().equals(x509Credential.getSubject()) && c.getIssuerDn().equals(x509Credential.getIssuer()))
-      .findAny();
+    // check if the x509Credential is linked to another user
+    if (linkedAccount.isPresent() && !linkedAccount.get().getUuid().equals(userAccount.getUuid())) {
+      throw new AccountAlreadyLinkedError(
+          format("X.509 credential with subject '%s' is already linked to another user",
+              x509Credential.getSubject()));
+    }
 
-    if (linkedCert.isPresent()) {
+    Optional<IamX509Certificate> linkedCertificate = certificateRepository
+      .findBySubjectAndIssuer(x509Credential.getSubject(), x509Credential.getIssuer());
 
-      linkedCert.ifPresent(c -> {
-        c.setCertificate(x509Credential.getCertificateChainPemString());
-        c.setLastUpdateTime(new Date());
-      });
+    if (linkedCertificate.isPresent()) {
 
+      linkedCertificate.get().setCertificate(x509Credential.getCertificateChainPemString());
+      linkedCertificate.get().setLastUpdateTime(new Date());
+      certificateRepository.save(linkedCertificate.get());
+      userAccount.getX509Certificates().add(linkedCertificate.get());
```

Comment:
Set.add insert the element only if is not contained and it uses equals method to check this. Unluckily our X509Certificate equals method checks only subject and issuer then no PEM certificate. This will mean that the element won't be replaced into that set. Probably it's not a big problem but it's better to remove and then add it again.